### PR TITLE
Starting the process of API discussion

### PR DESCRIPTION
**Pull request by rejuvyesh**
MAPOMDPs.jl should be the minimal API like package for defining multi-agent POMDP Models and their Solvers. This package will contain our factored value solvers as well as naive centralized solver using MCTS. We could potentially add decentralized MCTS here as well later.

If we look at the current [API](https://github.com/JuliaPOMDP/MAPOMDPs.jl/blob/master/src/MAPOMDPs.jl), we have vector of states/observations for each agent and a vector of actions for each agent in `gen` (to be changed to `@gen`). We also return a vector of rewards corresponding to each agent. So right now the setup can support POSG (partially observed stochastic game) if the rewards are different for each agent.

### CODE REVIEW

**Comment by Shushman** [comment]
I'll let the others chime in on the API as @rejuvyesh and I probably cut some corners for convenience in the paper experiments. I'm putting a mini-checklist of things to do and/or think about:

- [ ] Add docstrings to the various functions
- [ ] Should we call the centralized one FCMCTS? It's very close to FVMCTS and might cause confusion. I do realize we are using `JointMCTSTree` in `fv_mcts_vanilla`. So we could either change FCMCTS to JointMCTS and `JointMCTSTree` to `FVMCTSTree` or just change FCMCTS to MMCTS

File: src/MAMCTS.jl
Comment (Shushman) [inline src/MAMCTS.jl]: Perhaps `FactoredRandomPolicy` etc. can be in its own file?

File: src/fcmcts/fcmcts.jl
Comment (Shushman) [inline src/fcmcts/fcmcts.jl]: Note to self: Add docu ala https://github.com/JuliaPOMDP/MCTS.jl/blob/a24e92d108853a0410ac0a61f64d4e53a8623e71/src/vanilla.jl#L1-L52

File: src/fvmcts/fv_mcts_vanilla.jl
Comment (Shushman) [inline src/fvmcts/fv_mcts_vanilla.jl]: Note to self: Add docu ala https://github.com/JuliaPOMDP/MCTS.jl/blob/a24e92d108853a0410ac0a61f64d4e53a8623e71/src/vanilla.jl#L1-L52

File: src/fvmcts/fv_mcts_vanilla.jl
Comment (Shushman) [inline src/fvmcts/fv_mcts_vanilla.jl]: Critical docstring

**Comment by rejuvyesh** [comment]
> * Should we call the centralized one FCMCTS? It's very close to FVMCTS and might cause confusion. I do realize we are using `JointMCTSTree` in `fv_mcts_vanilla`. So we could either change FCMCTS to JointMCTS and `JointMCTSTree` to `FVMCTSTree` or just change FCMCTS to MMCTS

Agreed. Let's change `FCMCTS` to `JointMCTS` and current `JointMCTS*` to `FVMCTS*`

**Comment by zsunberg** [comment]
Nice - this seems cool - are you looking for feedback on this PR or on MAPOMDPs?

**Comment by zsunberg** [comment]
A few notes on MAPOMDPs:

- It should probably be called MultiAgentPOMDPs if we think we are ever going to register it.
- I think it should be `JointMDP{S, A<:AbstractVector} <: MDP{S, A}` because
  - the shared state might contain something more than just the individual states of the agents
  - it is good to let the problem writer specify the concrete action vector type otherwise MDP solvers might have trouble with type stability
- `POMDPs.reward` should return a scalar so that MDP/POMDP solvers can solve the problems. How does the problem writer specify the vector of rewards for each agent?
- Using `get` is not typical julia style. What is the reason you decided to use it?
- Docstrings do not normal contain `function`.

**Comment by Shushman** [comment]
Thanks for the notes @zsunberg ! Feedback on both would be great but the MAPOMDPs (or MultiAgentPOMDPs I guess) is more critical since it would be the problem template (versus this specific solver).

**Comment by rejuvyesh** [comment]
These are all good points.

> * It should probably be called MultiAgentPOMDPs if we think we are ever going to register it.

That's reasonable. Do you think we should rename this to MultiAgentMCTS too then.


> * `POMDPs.reward` should return a scalar so that MDP/POMDP solvers can solve the problems. How does the problem writer specify the vector of rewards for each agent?

Right now we were overloading this to return a vector of rewards. That's why we had to modify the rollout function for `JointMDP`. We'll have to think about this.

> * Using `get` is not typical julia style. What is the reason you decided to use it?

It was just easy. We'll rename these.

**Comment by zsunberg** [comment]
I think making `POMDPs.reward` return a scalar is pretty important. The point of making a `Joint(PO)MDP` a `(PO)MDP` is so that you can demonstrate that multi-agent approaches outperform monolithic POMDP solvers. If `POMDPs.reward` returns a vector, the monolithic POMDP solvers will not be able to solve it. I think you should introduce `MultiAgentPOMDPs.agent_rewards` and have
```julia
POMDPs.reward(m::JointMDP, s, a, sp) = sum(agent_rewards(m, s, a, sp))
```
by default.

**Comment by Shushman** [comment]
@rejuvyesh  from the `FVMCTSSolver` type (https://github.com/JuliaPOMDP/MAMCTS.jl/blob/impl/src/fvmcts/fv_mcts_vanilla.jl#L21), do we ever use the `reuse_tree` or `max_time` arguments anywhere? In the domains?

**Comment by rejuvyesh** [comment]
> @rejuvyesh from the `FVMCTSSolver` type (https://github.com/JuliaPOMDP/MAMCTS.jl/blob/impl/src/fvmcts/fv_mcts_vanilla.jl#L21), do we ever use the `reuse_tree` or `max_time` arguments anywhere? In the domains?

Not yet. But should do something about `max_time` at least.

File: src/fvmcts/fv_mcts_vanilla.jl
Comment (Shushman) [inline src/fvmcts/fv_mcts_vanilla.jl]: Change the `get_` style naming

File: src/fvmcts/fv_mcts_vanilla.jl
Comment (Shushman) [inline src/fvmcts/fv_mcts_vanilla.jl:321]: Is this what you meant by the new POMDPs.jl API? No more DDN??

File: src/fcmcts/fcmcts.jl
Comment (Shushman) [inline src/fcmcts/fcmcts.jl]: I see, so we're going to delegate the joint state / joint action aspect in FCMCTS to the domain?

Comment (Shushman) [reply src/fcmcts/fcmcts.jl]: Does beg the question of why we need FCMCTS at all. Can't we just use vanilla MCTS as the baseline then?

File: src/fcmcts/fcmcts.jl
Comment (Shushman) [inline src/fcmcts/fcmcts.jl]: As far as I can tell, this is the only place we are deviating from vanilla MCTS on an unfactored MMDP. I suppose it is worth maintaining separately for this alone

File: src/fcmcts/fcmcts.jl
Comment (Shushman) [inline src/fcmcts/fcmcts.jl]: Uncomment? Since we are using `::SV` further down?

File: src/fcmcts/fcmcts.jl
Comment (Shushman) [inline src/fcmcts/fcmcts.jl]: I think this is fine as we do use `enumerate(s)` above. We are implicitly requiring that the datastructure be something we can apply `eltype` to

File: src/fcmcts/fcmcts.jl
Comment (Shushman) [inline src/fcmcts/fcmcts.jl]: Wait, can we? Because we were doing `enumerate(s)` before when collecting the actions. Is `enumerate` possible on anything other than a vector?

File: src/fvmcts/action_coordination/varel.jl
Comment (Shushman) [inline src/fvmcts/action_coordination/varel.jl]: What's the issue here?

File: src/fvmcts/action_coordination/varel.jl
Comment (Shushman) [inline src/fvmcts/action_coordination/varel.jl]: I see. The TODO here is warranted haha. The equivalent for the `q::AbstractVector` case should be
`q_comp_value = q * length(comp)` ?

Comment (Shushman) [reply src/fvmcts/action_coordination/varel.jl]: Because if we had a global reward vector where all q values were the same, then the line `q_comp_value = sum(q[c] for c in comp)` would yield the same result right?

File: src/fvmcts/fv_mcts_vanilla.jl
Comment (Shushman) [inline src/fvmcts/fv_mcts_vanilla.jl:210]: Good catch! Even in the previous version, this should have been `S` and not `eltype`

**Review by Shushman — COMMENTED** [review]
Looks good! The only possible logical issue would have been the `update_statistics` where `q::Float64`, but I think my suggestion there should recover the same behavior.

File: src/fcmcts/fcmcts.jl
Comment (rejuvyesh) [reply src/fcmcts/fcmcts.jl]: I think we might be able to do away with `FCMCTS` need to check a few things with actions I think.

File: src/fcmcts/fcmcts.jl
Comment (rejuvyesh) [reply src/fcmcts/fcmcts.jl]: Right! But should be just define:

```julia
POMDPs.actions(mdp::JointMDP, s) =  vec(map(collect, Iterators.product((agent_actions(mdp, i, si) for (i, si) in enumerate(s))...)))
```

File: src/fcmcts/fcmcts.jl
Comment (rejuvyesh) [reply src/fcmcts/fcmcts.jl]: But `::SV` need not be `AbstractVector` and can have other global info too?

File: src/fcmcts/fcmcts.jl
Comment (rejuvyesh) [reply src/fcmcts/fcmcts.jl]: Yeah, this is something I am not sure about. Technically there can be some "global" information that is not provided to agents but then it becomes partially observed?

File: src/fvmcts/action_coordination/varel.jl
Comment (rejuvyesh) [reply src/fvmcts/action_coordination/varel.jl]: Yeah, that's the thing I wanted to confirm and seems alright.

**Comment by rejuvyesh** [comment]
@Shushman please take another look at the scaler global reward handling when you get the time.

**Comment by rejuvyesh** [comment]
@zsunberg Should be rename this to `MultiAgentMCTS.jl` as well?

File: src/fcmcts/fcmcts.jl
Comment (rejuvyesh) [reply src/fcmcts/fcmcts.jl]: Yeah, this is what I am unsure regarding @zsunberg's comments. We are assuming that `S` for JointMDP is something we can enumerate into agent specific things (which might just be copies of the same thing), but there could instead just be a single global state too.

File: src/fvmcts/fv_mcts_vanilla.jl
Comment (zsunberg) [reply src/fvmcts/fv_mcts_vanilla.jl:321]: Yes, as of POMDPs 0.9 we decided to get rid of the flexible DDN.

File: src/fcmcts/fcmcts.jl
Comment (zsunberg) [reply src/fcmcts/fcmcts.jl]: It makes sense to uncomment `SV` since it is used further down

File: src/fcmcts/fcmcts.jl
Comment (zsunberg) [reply src/fcmcts/fcmcts.jl]: I think this is fine. Worst case it tells the problem writer that they need to implement `agent_actions(::P, ::Int64, ::Any)`

File: src/fcmcts/fcmcts.jl
Comment (zsunberg) [reply src/fcmcts/fcmcts.jl]: I think `SV`, but see my main review comment.

**Review by zsunberg — APPROVED** [review]
The details that were commented on seem good to me, but @rejuvyesh 's comment

> Technically there can be some "global" information that is not provided to agents but then it becomes partially observed?

makes an interesting point. What is the definition of an agent state? in a JointMDP, do the agents get to make decisions based on `S` or `SV`?

**Comment by rejuvyesh** [comment]
> The details that were commented on seem good to me, but @rejuvyesh 's comment
> 
> > Technically there can be some "global" information that is not provided to agents but then it becomes partially observed?
> 
> makes an interesting point. What is the definition of an agent state? in a JointMDP, do the agents get to make decisions based on `S` or `SV`?

@Shushman , @mykelk correct me if I am wrong, but in MultiAgent MDP, all agents have access to the full global state information but the agent state could differ so as to allow say rotations or relative positions or agent identity?

**Comment by Shushman** [comment]
> > The details that were commented on seem good to me, but @rejuvyesh 's comment
> > > Technically there can be some "global" information that is not provided to agents but then it becomes partially observed?
> > 
> > 
> > makes an interesting point. What is the definition of an agent state? in a JointMDP, do the agents get to make decisions based on `S` or `SV`?
> 
> @Shushman , @mykelk correct me if I am wrong, but in MultiAgent MDP, all agents have access to the full global state information but the agent state could differ so as to allow say rotations or relative positions or agent identity?

According to DMU, in an MMDP, each agent is able to observe the full true state, i.e., the states of all other agents in the problem.
I don't think there needs to be a separate agent state besides the full global state? It just means that any 'agent' can see the full state vector at that timestep

**Comment by rejuvyesh** [comment]
> > > The details that were commented on seem good to me, but @rejuvyesh 's comment
> > > > Technically there can be some "global" information that is not provided to agents but then it becomes partially observed?
> > > 
> > > 
> > > makes an interesting point. What is the definition of an agent state? in a JointMDP, do the agents get to make decisions based on `S` or `SV`?
> > 
> > 
> > @Shushman , @mykelk correct me if I am wrong, but in MultiAgent MDP, all agents have access to the full global state information but the agent state could differ so as to allow say rotations or relative positions or agent identity?
> 
> According to DMU, in an MMDP, each agent is able to observe the full true state, i.e., the states of all other agents in the problem.
> I don't think there needs to be a separate agent state besides the full global state? It just means that any 'agent' can see the full state vector at that timestep

Hmm. But then the issue for us is that we shouldn't require `enumerate` on `JointMDP` state then, because JuliaPOMDP philosophy is to not restrict the types unless required and the full state need not be a vector (there could just be a global view for example).

**Comment by zsunberg** [comment]
> According to DMU, in an MMDP, each agent is able to observe the full true state, i.e., the states of all other agents in the problem. I don't think there needs to be a separate agent state besides the full global state? It just means that any 'agent' can see the full state vector at that timestep

Ah, OK, this makes sense to me. Whenever an agent needs to make a decision, it should be given access to the global state, and there is no reason the global state needs to be an `AbstractVector`.

Then is there a reason to have any concept of individual states?

Perhaps individual states would be useful for the common *suboptimal* solution of having agents feedback on their own states or a smaller view of the state? If the concept of individual states is needed, I think it would be reasonable to say that states in a `JointMDP` should support the [iteration interface](https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-iteration) and/or have `getindex`. `Vector` and `Tuple` are types that supports both of these. Or it may be better to have a separate `agentview` function for this.

File: src/fcmcts/fcmcts.jl
Comment (Shushman) [reply src/fcmcts/fcmcts.jl]: Actually, yeah if we do this then we shouldn't need FCMCTS anymore at all!
What we can do for now is to remove `FCMCTS` from this repo so it does not hold up the rest of it. Later on, we can fix our own baseline scripts to use vanilla MCTS.

File: src/MAMCTS.jl
Comment (Shushman) [inline src/MAMCTS.jl]: Eh under the circumstances it is the least bad option, I think.

File: src/fvmcts/fv_mcts_vanilla.jl
Comment (Shushman) [inline src/fvmcts/fv_mcts_vanilla.jl]: Not sure I see the issue here. Don't we require a generative model to be implemented?

**Review by Shushman — COMMENTED** [review]
Scalar global rewards (and the others) look good to me

**Comment by rejuvyesh** [comment]
Renamed to FactoredValueMCTS.jl everywhere. Will figure out later if `FCMCTS` is even required.

File: src/MAMCTS.jl
Comment (rejuvyesh) [reply src/MAMCTS.jl]: I mean, I could do a random action as well?

File: src/fvmcts/fv_mcts_vanilla.jl
Comment (rejuvyesh) [reply src/fvmcts/fv_mcts_vanilla.jl]: MCTS.jl has the same thing 😛. @zsunberg might know the exact reason for the comment.

**Pull request merged**
